services/wfm/src/main/java/org/openkilda/wfm/topology/flow/transactions/FlowCommandRegistry.java

```diff
@@ -0,0 +1,252 @@
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.wfm.topology.flow.transactions;
+
+import static java.lang.String.format;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.unmodifiableList;
+import static java.util.stream.Collectors.toMap;
+
+import org.openkilda.messaging.command.flow.BaseFlow;
+
+import com.fasterxml.uuid.EthernetAddress;
+import com.fasterxml.uuid.Generators;
+import com.fasterxml.uuid.impl.TimeBasedGenerator;
+import lombok.extern.slf4j.Slf4j;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class FlowCommandRegistry {
+    private final TimeBasedGenerator batchIdGenerator =
+            Generators.timeBasedGenerator(EthernetAddress.fromInterface());
+
+    private final Map<String, Queue<Group>> groups = new HashMap<>();
+    private final Map<UUID, Batch> batches = new HashMap<>();
+    private final Map<UUID, UUID> transactionToBatch = new HashMap<>();
+
+    /**
+     * Registers the grouped commands as a batch for the flow.
+     * Puts the commands to the back of the queue.
+     */
+    public synchronized void registerBatch(String flowId, List<List<BaseFlow>> batchCommands) {
+        UUID batchId = batchIdGenerator.generate();
+        Set<UUID> batchTransactions = new HashSet<>();
+
+        log.debug("Registering commands as batch {}: {}", batchId, batchCommands);
+
+        Queue<Group> flowGroups = groups.computeIfAbsent(flowId, k -> new LinkedList<>());
+        for (List<BaseFlow> groupCommands : batchCommands) {
+            flowGroups.add(new Group(batchId, new ArrayList<>(groupCommands)));
+
+            for (BaseFlow command : groupCommands) {
+                if (!command.getId().equals(flowId)) {
+                    throw new IllegalArgumentException(
+                            format("Command '%s' doesn't correspond to specified flow %s", command, flowId));
+                }
+
+                UUID transactionId = command.getTransactionId();
+                batchTransactions.add(transactionId);
+                transactionToBatch.put(transactionId, batchId);
+            }
+        }
+
+        batches.put(batchId, new Batch(flowId, batchTransactions));
+    }
+
+    /**
+     * Removes the command (identified by the flow and transaction IDs) from the current group.
+     * Cleans up groups and batches if they are empty.
+     */
+    public synchronized void removeCommand(String flowId, UUID transactionId) throws UnknownTransactionException {
```

Comment:
`synchnozized` in storm bolts? Why for?

Reply:
Good point. Removed and added notice to the class javadoc.